### PR TITLE
dnn: add openvino, opencl and cuda backends for layer normalization layer

### PR DESCRIPTION
**Pull request by fengyuentau**
Merge after https://github.com/opencv/opencv/pull/24544.

Todo:

- [x] openvino
- [x] opencl
- [x] cuda

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Linux OpenCL,Win64 OpenCL
```

### CODE REVIEW

**Comment by dkurt** [comment]
Oh, wait a minute, why we need continue development of LayerNorm layer while there is a new InstanceNorm layer https://github.com/opencv/opencv/pull/24378? I expected that LayerNorm should be removed completely.

**Comment by fengyuentau** [comment]
We agreed previously in the meeting that we keep two classes, and they share similar kernels for CPU.

File: modules/dnn/src/opencl/mvn.cl [hunk 1]
```diff
@@ -126,12 +126,18 @@ __kernel void MVN(__global const Dtype* src,
...
 #ifdef FUSE_BATCH_NORM
     w = bnorm_weight[x % channels];
     b = bnorm_bias[x % channels];
 #endif
 
+#endif // LAYER_NORM
+
     vec_type src_vec = load(src, index) - (vec_type)mean_val;
     vec_type dst_vec = src_vec * alpha;
     dst_vec = dst_vec * w + (vec_type)b;
```
Comment (dkurt) [inline modules/dnn/src/opencl/mvn.cl:143]:
```suggestion
    vec_type w = 1.f, b = 0.f;
#ifdef FUSE_BATCH_NORM
    w = bnorm_weight[x % channels];
    b = bnorm_bias[x % channels];
#endif

#endif // LAYER_NORM

    vec_type src_vec = load(src, index) - (vec_type)mean_val;
    vec_type dst_vec = src_vec * alpha;
    dst_vec = dst_vec * w + b;
```

**Comment by dkurt** [comment]
Let's see the final state. As far as I remember the only difference between them is scalar or vector scale/mean value,

**Comment by fengyuentau** [comment]
> Let's see the final state. As far as I remember the only difference between them is scalar or vector scale/bias values,

Indexing is also different as well. For Instance Normalization, its indexing is like `scale[outer_idx % C]`; while for Layer Normalization, its indexing is like `scale[j] for j in inner_size`.

File: modules/dnn/src/layers/layer_norm.cpp [hunk 2]
```diff
@@ -25,6 +43,10 @@ class LayerNormLayerImpl CV_FINAL : public LayerNormLayer
 
     virtual bool supportBackend(int backendId) CV_OVERRIDE
     {
+#ifdef HAVE_INF_ENGINE
+        if (backendId == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
+            return true;
+#endif
         return backendId == DNN_BACKEND_OPENCV ||
                (backendId == DNN_BACKEND_CANN && axis != -1); // axis=-1 not supported due to 1d mat shape problem
     }
```
Comment (dkurt) [inline modules/dnn/src/layers/layer_norm.cpp:52]: there is no check for CUDA

File: modules/dnn/src/cuda/mvn.cu [hunk 3]
```diff
@@ -77,6 +77,27 @@ namespace raw {
             output[idx] = (static_cast<float>(input[idx]) - means[outer_idx]) * s + b;
         }
     }
+
+    template <class T>
+    __global__ void normalize_mean_variance_layernorm(Span<T> output, View<T> input, View<T> scale, View<float> means, View<float> stdev, size_type inner_size) {
+        for (auto idx : grid_stride_range(output.size())) {
+            const index_type outer_idx = idx / inner_size;
+            const index_type inner_idx = idx % inner_size;
+            auto s = static_cast<float>(scale[inner_idx]) * stdev[outer_idx];
```
Comment (dkurt) [inline modules/dnn/src/cuda/mvn.cu:86]: `inv_stddev`?

Comment (fengyuentau) [reply modules/dnn/src/cuda/mvn.cu:86]: it is already inverted in the above kernel `compute_normalization_scale`.

File: modules/dnn/src/layers/layer_norm.cpp (hunk 2 above)
Comment (fengyuentau) [reply modules/dnn/src/layers/layer_norm.cpp:53]: yes, thanks

File: modules/dnn/src/cuda/mvn.cu (hunk 3 above)
Comment (dkurt) [reply modules/dnn/src/cuda/mvn.cu:86]: I mean just a name for a variable. It's expected to multiply on an inverse deviation.

Comment (fengyuentau) [reply modules/dnn/src/cuda/mvn.cu]: Ok.

**Review by dkurt — APPROVED** [review]

**Pull request merged**